src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,231 @@
+  bp_cache::BeliefPropagationCache,
+  partition_vertices::Vector{<:PartitionVertex};
+  ignore_edges=PartitionEdge[],
+)
+  bpes = boundary_partitionedges(bp_cache, partition_vertices; dir=:in)
+  ms = messages(bp_cache, setdiff(bpes, ignore_edges))
+  return reduce(vcat, ms; init=[])
+end
+
+function incoming_messages(
+  bp_cache::BeliefPropagationCache, partition_vertex::PartitionVertex; kwargs...
+)
+  return incoming_messages(bp_cache, [partition_vertex]; kwargs...)
+end
+
+function incoming_messages(bp_cache::BeliefPropagationCache, verts::Vector)
+  partition_verts = partitionvertices(bp_cache, verts)
+  messages = incoming_messages(bp_cache, partition_verts)
+  central_tensors = ITensor[
+    tensornetwork(bp_cache)[v] for v in setdiff(vertices(bp_cache, partition_verts), verts)
+  ]
+  return vcat(messages, central_tensors)
+end
+
+function incoming_messages(bp_cache::BeliefPropagationCache, vertex)
+  return incoming_messages(bp_cache, [vertex])
+end
+
+function factor(bp_cache::BeliefPropagationCache, vertex::PartitionVertex)
+  ptn = partitioned_itensornetwork(bp_cache)
+  return Vector{ITensor}(subgraph(ptn, vertex))
+end
+
+"""
+Compute message tensor as product of incoming mts and local state
+"""
+function update_message(
+  bp_cache::BeliefPropagationCache,
+  edge::PartitionEdge;
+  message_update=default_message_update,
+  message_update_kwargs=default_message_update_kwargs(),
+)
+  vertex = src(edge)
+  messages = incoming_messages(bp_cache, vertex; ignore_edges=PartitionEdge[reverse(edge)])
+  state = factor(bp_cache, vertex)
+
+  return message_update(ITensor[messages; state]; message_update_kwargs...)
+end
+
+"""
+Do a sequential update of the message tensors on `edges`
+"""
+function update(
+  bp_cache::BeliefPropagationCache,
+  edges::Vector{<:PartitionEdge};
+  (update_diff!)=nothing,
+  kwargs...,
+)
+  bp_cache_updated = copy(bp_cache)
+  mts = messages(bp_cache_updated)
+  for e in edges
+    set!(mts, e, update_message(bp_cache_updated, e; kwargs...))
+    if !isnothing(update_diff!)
+      update_diff![] += message_diff(message(bp_cache, e), mts[e])
+    end
+  end
+  return bp_cache_updated
+end
+
+"""
+Update the message tensor on a single edge
+"""
+function update(bp_cache::BeliefPropagationCache, edge::PartitionEdge; kwargs...)
+  return update(bp_cache, [edge]; kwargs...)
+end
+
+"""
+Do parallel updates between groups of edges of all message tensors
+Currently we send the full message tensor data struct to update for each edge_group. But really we only need the
+mts relevant to that group.
+"""
+function update(
+  bp_cache::BeliefPropagationCache,
+  edge_groups::Vector{<:Vector{<:PartitionEdge}};
+  kwargs...,
+)
+  new_mts = copy(messages(bp_cache))
+  for edges in edge_groups
+    bp_cache_t = update(bp_cache, edges; kwargs...)
+    for e in edges
+      new_mts[e] = message(bp_cache_t, e)
+    end
+  end
+  return BeliefPropagationCache(copy(partitioned_itensornetwork(bp_cache)), new_mts)
```

Comment:
We could define `set_messages(cache::BeliefPropagationCache, messages) = BeliefPropagationCache(partitioned_itensornetwork(cache), messages, default_message(cache))` as a shorthand for this. In general, I like defining getters and setters for fields of types since it is pretty common to want to update just one field and keep the rest the same. Also, [Accessors.jl](https://github.com/JuliaObjects/Accessors.jl) provides nice syntax for implementing field setters.

Reply:
That makes sense. I have now defined that.